man/ndb_perror.1
'\" t
.\"     Title: ndb_perror
.\"    Author: [FIXME: author] [see http://docbook.sf.net/el/author]
.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
.\"      Date: 06/25/2026
.\"    Manual: MySQL Database System
.\"    Source: MySQL 9.7
.\"  Language: English
.\"
.TH "NDB_PERROR" "1" "06/25/2026" "MySQL 9\&.7" "MySQL Database System"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
ndb_perror \- obtain NDB error message information
.SH "SYNOPSIS"
.HP \w'\fBndb_perror\ \fR\fB\fIoptions\fR\fR\ 'u
\fBndb_perror \fR\fB\fIoptions\fR\fR
.SH "DESCRIPTION"
.PP
\fBndb_perror\fR
shows information about an NDB error, given its error code\&. This includes the error message, the type of error, and whether the error is permanent or temporary\&. This is intended as a drop\-in replacement for
\fBperror\fR
\fB\-\-ndb\fR, which is no longer supported\&.
Usage
.sp
.if n \{\
.RS 4
.\}
.nf
ndb_perror [\fIoptions\fR] \fIerror_code\fR
.fi
.if n \{\
.RE
.\}
.PP
\fBndb_perror\fR
does not need to access a running NDB Cluster, or any nodes (including SQL nodes)\&. To view information about a given NDB error, invoke the program, using the error code as an argument, like this:
.sp
.if n \{\
.RS 4
.\}
.nf
$> \fBndb_perror 323\fR
NDB error code 323: Invalid nodegroup id, nodegroup already existing: Permanent error: Application error
.fi
.if n \{\
.RE
.\}
.PP
To display only the error message, invoke
\fBndb_perror\fR
with the
\fB\-\-silent\fR
option (short form
\fB\-s\fR), as shown here:
.sp
.if n \{\
.RS 4
.\}
.nf
$> \fBndb_perror \-s 323\fR
Invalid nodegroup id, nodegroup already existing: Permanent error: Application error
.fi
.if n \{\
.RE
.\}
.PP
Like
\fBperror\fR,
\fBndb_perror\fR
accepts multiple error codes:
.sp
.if n \{\
.RS 4
.\}
.nf
$> \fBndb_perror 321 1001\fR
NDB error code 321: Invalid nodegroup id: Permanent error: Application error
NDB error code 1001: Illegal connect string
.fi
.if n \{\
.RE
.\}
.PP
Additional program options for
\fBndb_perror\fR
are described later in this section\&.
.PP
\fBndb_perror\fR
replaces
\fBperror\fR
\fB\-\-ndb\fR, which is no longer supported by NDB Cluster\&. To make substitution easier in scripts and other applications that might depend on
\fBperror\fR
for obtaining NDB error information,
\fBndb_perror\fR
supports its own
\(lqdummy\(rq
\fB\-\-ndb\fR
option, which does nothing\&.
.PP
The following table includes all options that are specific to the NDB Cluster program
\fBndb_perror\fR\&. Additional descriptions follow the table\&.
.PP
Additional Options
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-defaults\-extra\-file\fR
.TS
allbox tab(:);
lB l
lB l
lB l.
T{
Command-Line Format
T}:T{
--defaults-extra-file=path
T}
T{
Type
T}:T{
String
T}
T{
Default Value
T}:T{
[none]
T}
.TE
.sp 1
Read given file after global files are read\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-defaults\-file\fR
.TS
allbox tab(:);
lB l
lB l
lB l.
T{
Command-Line Format
T}:T{
--defaults-file=path
T}
T{
Type
T}:T{
String
T}
T{
Default Value
T}:T{
[none]
T}
.TE
.sp 1
Read default options from given file only\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-defaults\-group\-suffix\fR
.TS
allbox tab(:);
lB l
lB l
lB l.
T{
Command-Line Format
T}:T{
--defaults-group-suffix=string
T}
T{
Type
T}:T{
String
T}
T{
Default Value
T}:T{
[none]
T}
.TE
.sp 1
Also read groups with concat(group, suffix)\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-help\fR,
\fB\-?\fR
.TS
allbox tab(:);
lB l.
T{
Command-Line Format
T}:T{
--help
T}
.TE
.sp 1
Display program help text and exit\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-login\-path\fR
.TS
allbox tab(:);
lB l
lB l
lB l.
T{
Command-Line Format
T}:T{
--login-path=path
T}
T{
Type
T}:T{
String
T}
T{
Default Value
T}:T{
[none]
T}
.TE
.sp 1
Read given path from login file\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-no\-login\-paths\fR
.TS
allbox tab(:);
lB l.
T{
Command-Line Format
T}:T{
--no-login-paths
T}
.TE
.sp 1
Skips reading options from the login path file\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-ndb\fR
.TS
allbox tab(:);
lB l.
T{
Command-Line Format
T}:T{
--ndb
T}
.TE
.sp 1
For compatibility with applications depending on old versions of
\fBperror\fR
that use that program\*(Aqs
\fB\-\-ndb\fR
option\&. The option when used with
\fBndb_perror\fR
does nothing, and is ignored by it\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-no\-defaults\fR
.TS
allbox tab(:);
lB l.
T{
Command-Line Format
T}:T{
--no-defaults
T}
.TE
.sp 1
Do not read default options from any option file other than login file\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-print\-defaults\fR
.TS
allbox tab(:);
lB l.
T{
Command-Line Format
T}:T{
--print-defaults
T}
.TE
.sp 1
Print program argument list and exit\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-silent\fR,
\fB\-s\fR
.TS
allbox tab(:);
lB l.
T{
Command-Line Format
T}:T{
--silent
T}
.TE
.sp 1
Show error message only\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-version\fR,
\fB\-V\fR
.TS
allbox tab(:);
lB l.
T{
Command-Line Format
T}:T{
--version
T}
.TE
.sp 1
Print program version information and exit\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-\-verbose\fR,
\fB\-v\fR
.TS
allbox tab(:);
lB l.
T{
Command-Line Format
T}:T{
--verbose
T}
.TE
.sp 1
Verbose output; disable with
\fB\-\-silent\fR\&.
.RE
.SH "COPYRIGHT"
.br
.PP
Copyright \(co 1997, 2026, Oracle and/or its affiliates.
.PP
This documentation is free software; you can redistribute it and/or modify it only under the terms of the GNU General Public License as published by the Free Software Foundation; version 2 of the License.
.PP
This documentation is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
.PP
You should have received a copy of the GNU General Public License along with the program; if not, write to the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA or see http://www.gnu.org/licenses/.
.sp
.SH "SEE ALSO"
For more information, please refer to the MySQL Reference Manual,
which may already be installed locally and which is also available
online at http://dev.mysql.com/doc/.
.SH AUTHOR
Oracle Corporation (http://dev.mysql.com/).


ChangeLog
2010-03-18	The InnoDB Team

	* CMakeLists.txt:
	Fix Bug#52102 InnoDB Plugin shows performance drop compared to
	InnoDB (Windows)

2010-03-18	The InnoDB Team

	* buf0buf.ic:
	When comparing the time of the first access to a block against
	innodb_old_blocks_time, use 32-bit arithmetics. The comparison was
	incorrect on 64-bit systems.

2010-03-11	The InnoDB Team

	* buf0buf.h, buf0buf.ic:
	Fix and clarify the latching of some buf_block_t members.
	Note that check_index_page_at_flush is not protected by any mutex.
	Note and assert that lock_hash_val is protected by the rw-latch.

2010-03-10	The InnoDB Team

	* trx/trx0sys.c:
	Fix Bug#51653 outdated reference to set-variable

2010-03-10	The InnoDB Team

	* handler/ha_innodb.cc, mysql-test/innodb_bug21704.result,
	mysql-test/innodb_bug47621.result, mysql-test/innodb_bug47621.test:
	Fix Bug#47621 MySQL and InnoDB data dictionaries will become out of
	sync when renaming columns

2010-03-10	The InnoDB Team

	* handler/ha_innodb.cc:
	Fix Bug#51356 Many Valgrind errors in error messages
	with concurrent DDL

2010-03-10	The InnoDB Team

	* handler/ha_innodb.cc, handler/handler0alter.cc,
	mysql-test/innodb_bug51378.result, mysql-test/innodb_bug51378.test:
	Fix Bug#51378 Init 'ref_length' to correct value, in case an out
	of bound MySQL primary_key

2010-03-10	The InnoDB Team

	* log/log0recv.c:
	Remove a bogus assertion about page numbers exceeding 0x90000000
	in the redo log. Abort when encountering a corrupted redo log
	record, unless innodb_force_recovery is set.

2010-03-09	The InnoDB Team

	* handler/ha_innodb.cc:
	Make SHOW ENGINE INNODB MUTEX STATUS display SUM(os_waits)
	for the buffer pool block mutexes and locks.

2010-03-08	The InnoDB Team

	* fil/fil0fil.c:
	Fix ALTER TABLE ... IMPORT TABLESPACE of compressed tables.

2010-03-03	The InnoDB Team

	* handler/handler0alter.cc, innodb-index.result, innodb-index.test,
	innodb.result, innodb.test:
	Disallow a duplicate index name when creating an index.

2010-02-11	The InnoDB Team

	* include/mem0mem.h, include/mem0mem.ic, mem/mem0mem.c:
	Fix Bug#49535 Available memory check slows down crash
	recovery tens of times

2010-02-09	The InnoDB Team

	* buf/buf0buf.c:
	Fix Bug#38901 InnoDB logs error repeatedly when trying to load
	page into buffer pool

2010-02-09	The InnoDB Team

	* srv/srv0srv.c:
	Let the master thread sleep if the amount of work to be done is
	calibrated as taking less than a second.

2010-02-04	The InnoDB Team

	* btr/btr0btr.c, btr/btr0cur.c, btr/btr0pcur.c, buf/buf0buf.c,
	include/btr0btr.h, include/btr0cur.h, include/btr0pcur.h,
	include/btr0pcur.ic, include/buf0buf.h, row/row0ins.c, row/row0sel.c:
	Pass the file name and line number of the caller of the
	b-tree cursor functions to the buffer pool requests, in order
	to make the latch diagnostics more accurate.

2010-02-03	The InnoDB Team

	* lock/lock0lock.c:
	Fix Bug#49001 SHOW INNODB STATUS deadlock info incorrect
	when deadlock detection aborts

2010-02-03	The InnoDB Team

	* buf/buf0lru.c:
	Fix Bug#35077 Very slow DROP TABLE (ALTER TABLE, OPTIMIZE TABLE)
	on compressed tables

2010-02-03	The InnoDB Team

	* handler/ha_innodb.cc, include/row0mysql.h, row/row0mysql.c:
	Clean up CHECK TABLE error handling.

2010-02-01	The InnoDB Team

	* handler/ha_innodb.cc, mysql-test/innodb-autoinc.test,
	mysql-test/innodb-autoinc.result,
	mysql-test/innodb-autoinc-44030.test,
	mysql-test/innodb-autoinc-44030.result:
	Fix Bug#49497 Error 1467 (ER_AUTOINC_READ_FAILED) on inserting
	a negative value

2010-01-27	The InnoDB Team

	* include/row0mysql.h, log/log0recv.c, row/row0mysql.c:
	Drop temporary tables at startup.
	This addresses the third aspect of
	Bug#41609 Crash recovery does not work for InnoDB temporary tables.

2010-01-21	The InnoDB Team

	* buf/buf0buf.c:
	Do not merge buffered inserts to compressed pages before
	the redo log has been applied in crash recovery.

2010-01-13	The InnoDB Team

	* row/row0sel.c:
	On the READ UNCOMMITTED isolation level, do not attempt to access
	a clustered index record that has been marked for deletion. The
	built-in InnoDB in MySQL 5.1 and earlier would attempt to retrieve
	a previous version of the record in this case.

2010-01-13	The InnoDB Team

	* buf/buf0buf.c:
	When disabling the adaptive hash index, check the block state
	before checking block->is_hashed, because the latter may be
	uninitialized right after server startup.

2010-01-12	The InnoDB Team

	* handler/ha_innodb.cc, handler/ha_innodb.h:
	Fix Bug#46193 crash when accessing tables after enabling
	innodb_force_recovery option

2010-01-12	The InnoDB Team

	* row/row0mysql.c:
	Fix Bug#49238 Creating/Dropping a temporary table while at 1023
	transactions will cause assert.

2009-12-02	The InnoDB Team

	* srv/srv0start.c:
	Display the zlib version number at startup.
	InnoDB compressed tables use zlib, and the implementation depends
	on the zlib function compressBound(), whose definition was slightly
	changed in zlib version 1.2.3.1 in 2006. MySQL bundles zlib 1.2.3
	from 2005, but some installations use a more recent zlib.

2009-11-30	The InnoDB Team

	* dict/dict0crea.c, dict/dict0mem.c, dict/dict0load.c,
	dict/dict0boot.c, fil/fil0fil.c, handler/ha_innodb.cc,
	include/dict0mem.h, row/row0mysql.c:
	Fix the bogus warning messages for non-existing temporary
	tables that were reported in
	Bug#41609 Crash recovery does not work for InnoDB temporary tables.
	The actual crash recovery bug was corrected on 2009-04-29.

2009-11-27	The InnoDB Team

	InnoDB Plugin 1.0.6 released

2009-11-20	The InnoDB Team

	* handler/ha_innodb.cc:
	Add a workaround to prevent a crash due to Bug#45961 DDL on
	partitioned innodb tables leaves data dictionary in an inconsistent
	state

2009-11-19	The InnoDB Team

	* btr/btr0btr.c:
	Fix Bug#48469 when innodb tablespace is configured too small, crash
	and corruption!

2009-11-19	The InnoDB Team

	* data/data0type.c:
	Fix Bug#48526 Data type for float and double is incorrectly reported
	in InnoDB table monitor


storage/ndb/tools/ndb_perror.cpp
/*
   Copyright (c) 2017, 2026, Oracle and/or its affiliates.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License, version 2.0,
   as published by the Free Software Foundation.

   This program is designed to work with certain software (including
   but not limited to OpenSSL) that is licensed under separate terms,
   as designated in a particular file or component or in included license
   documentation.  The authors of MySQL hereby grant you an additional
   permission to link the program and your derivative works with the
   separately licensed software that they have either included with
   the program or referenced in the documentation.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License, version 2.0, for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301  USA */

/*
  Return error-text for NDB error messages in same
  fashion as "perror --ndb <error>"
*/

#include "ndb_global.h"
#include "ndb_opts.h"

#include "mgmapi/mgmapi_error.h"
#include "mgmapi/ndbd_exit_codes.h"
#include "my_alloc.h"
#include "ndbapi/ndberror.h"

static bool opt_verbose;
static bool opt_silent;  // Overrides verbose and sets it to 0

static struct my_option my_long_options[] = {
    NdbStdOpt::help,
    {"ndb", NDB_OPT_NOSHORT,
     "For command line compatibility with 'perror --ndb', ignored.", nullptr,
     nullptr, nullptr, GET_NO_ARG, NO_ARG, 0, 0, 0, nullptr, 0, nullptr},
    {"silent", 's', "Only print the error message.", &opt_silent, nullptr,
     nullptr, GET_BOOL, NO_ARG, 0, 0, 0, nullptr, 0, nullptr},
    {"verbose", 'v', "Print error code and message (default).", &opt_verbose,
     nullptr, nullptr, GET_BOOL, NO_ARG, 1, 0, 0, nullptr, 0, nullptr},
    NdbStdOpt::version,
    NdbStdOpt::end_of_options};

const char *load_default_groups[] = {nullptr};

static void short_usage_sub() {
  ndb_short_usage_sub("[ERRORCODE [ERRORCODE...]]");
}

static int mgmapi_error_string(int err_no, char *str, int size) {
  for (int i = 0; i < ndb_mgm_noOfErrorMsgs; i++) {
    if ((int)ndb_mgm_error_msgs[i].code == err_no) {
      snprintf(str, size - 1, "%s", ndb_mgm_error_msgs[i].msg);
      str[size - 1] = '\0';
      return 1;  // Found a message
    }
  }
  return -1;
}

// Forward declare function from ndbd_exit_codes.cc which is not
// declared in any header
int ndbd_exit_string(int err_no, char *str, unsigned int size);

int main(int argc, char **argv) {
  NDB_INIT(argv[0]);
  Ndb_opts opts(argc, argv, my_long_options, load_default_groups);
  opts.set_usage_funcs(short_usage_sub);

  if (opts.handle_options() != 0) exit(255);

  if (opt_silent) {
    // --silent overrides any verbose setting
    opt_verbose = false;
  }

  if (!argc) {
    opts.usage();
    exit(1);
  }

  int error = 0;
  for (; argc-- > 0; argv++) {
    int code = atoi(*argv);

    char error_string[1024];
    if ((ndb_error_string(code, error_string, sizeof(error_string)) > 0) ||
        (ndbd_exit_string(code, error_string, sizeof(error_string)) > 0) ||
        (mgmapi_error_string(code, error_string, sizeof(error_string)) > 0)) {
      if (opt_verbose)
        printf("NDB error code %3d: %s\n", code, error_string);
      else
        puts(error_string);
    } else {
      fprintf(stderr, "Illegal ndb error code: %d\n", code);
      error = 1;
    }
  }

  exit(error);
  return error;
}
